### ISSUE TITLE
Include statements

### ISSUE THREAD

**Issue by shelbyd** [pureconfig/pureconfig#14]
I think we want to be careful with include statements.

## Problem

It's common to obscure where symbols are coming from. Consider the following file:

```
#include foo.pure
#include bar.pure

baz => qux
  quux = "something"
```

Where does the `qux` symbol come from? If we do globs as suggested in #2, it get worse.
What if both foo and bar have `qux` defined?

## Solution

I propose we use a syntax similar to [TypeScript](https://www.typescriptlang.org/docs/handbook/module-resolution.html). With this, the previous file would look like the following:

```
import {qux} from './foo.pure'
import {qux as unused} from './bar.pure'
```

Now it's clear where each symbol is coming from. And if multiple files have the same symbol, one can alias the symbol in the current file.

**Comment by pureconfig** [comment]
I share the concern, but would still like a super-simple syntax to just include everything. After all, these are config files. I think the main reason to use includes is to inject a template. Maybe keep it dirt simple in the first version, then support symbol-specific imports later on, if we find that it is needed?

The "from" keyword is superfluous and include is better than import in a file based system like this.

Does this work for you?

```
%include template.pure
%include {qux} foo.pure
%include {qux as unused} bar.pure
```

?

**Comment by shelbyd** [comment]
The syntax you gave is good.

I'd want full imports to be namespaced so we don't have collisions. One of the following:

```
%include template template.pure
%include {* as template} template.pure
%include * as template template.pure
```

I prefer the first one.
The second may be ambiguous with importing individual symbols.
The third is just another option.

Also, I think we'd want to specify about where the imported files come from. They could implicitly be relative to the current file, or relative to $PWD. We should decide either way at some point.

**Comment by Phlosioneer** [comment]
Another option is to treat everything included as though it were in a group. Then we can use existing syntax and parsing to work with it.

The problem config file would become

```
#include foo.pure
#include bar.pure

baz => foo.qux
  quux = "something"
```

This makes it crystal clear where symbols are coming from, without complicating the include directives. If a value is being used super often, the config file can just use an existing language construct to "import" it:
```
#include foo.pure
#include bar.pure

# "Using" statements, that are really just ordinary config statements:
qux => foo.qux

# qux is now defined in this file, and so can be used without qualification.
baz => qux
  quux = "something"
```
This allows the complexity of importing and using to be "opt-in", while also minimizing the special cases. We'll have to do something about globs, though.

**Comment by shelbyd** [comment]
I don't think we want to default to a namespace based on the name of the included file.

How do we deal with files in subdirectories? My best idea is to have each directory dot separated as a namespace. But that opens up ambiguity between `foo/bar.qux.pure` and `foo.bar/qux.pure`. Also, how do we deal with non `.pure` file endings? What is the namespace for `%include foo.bar`? Automatically making names might be slightly easier, but it remove clarity and complicates something that is supposed to be simple.

It's very simple if the user always has to specify the namespace for the imported file. I think the clearest syntax would probably be `include foo.pure as bar`. We technically wouldn't need the 'as', but it adds clarity and disambiguates from running multiple imports.

**Comment by pureconfig** [comment]
@shelbyd @Phlosioneer  I'm starting to wonder if the "name your config files whatever" was a bad idea. 

By requiring a *pure* extension, it'll be easier to write editor plugins, people will know what file type it is and, for the topic at hand, simplify include matters.

**Comment by shelbyd** [comment]
AFAIK no language actually forces you to use a specific extension. Most make it easier if you choose their specific extension, but you can work hard to use your own extension.

It seems we're fighting hard and making decisions with strange consequences to avoid two words. I think we should strive for clarity and simplicity without sacrificing power, and anything other than specifying a namespace for an included file is either unclear or very complex.

**Comment by dansouza** [comment]
> Where does the qux symbol come from? If we do globs as suggested in #2, it get worse.
> What if both foo and bar have qux defined?

I think it would be better (and simpler) to just default to whatever was defined last. So if a symbol is defined on 'foo.pure' and 'bar.pure', whatever file was parsed last has the final value for the symbol.

In case of globs, globs should always be parsed alphabetically, so if a symbol with the same name is defined in `conf.d/foo.pure` and `conf.d/bar.pure`, and you `%include conf.d/*.pure`, the value from `conf.d/bar.pure` will be the actual value for the symbol because it was included last.

That's how pretty much everything works like `bash` , etc.

**Comment by shelbyd** [comment]
We can certainly come up with a consistent ruleset for global includes. My point is that while it saves a little time when writing the config, it costs a huge amount of time when reading the config.

For the example I gave in the first comment, the problem is not for the system, it's for the user. The user has to look through the files to find the symbol they're looking for. And if they start from the top because they don't know the rules and find a symbol that's being overridden by a later file, that's a huge frustration. I have quite a few more examples where global includes cause problems, please ask if you're not convinced. :smile:

I expect that if we have global includes, the convention will be to namespace their config files to avoid polluting the global namespace. Which just makes more work when writing and reading.

What are the arguments against requiring a namespace for all includes? It takes more characters when writing and reading. But I think that is dramatically outweighed by the clarity it adds.

**Comment by pureconfig** [comment]
> What are the arguments against requiring a namespace for all includes? It takes more characters when writing and reading

The very reason I started Pure was to make it super obvious to edit for all kinds of people. However, includes will be rare and only programmers/devops will ever mock around with includes. So let's make includes robust.

@shelbyd In short, I agree with your namespaced include proposal. Let's go for that.

**Comment by pureconfig** [comment]
> Also, I think we'd want to specify about where the imported files come from. They could implicitly be relative to the current file, or relative to $PWD. We should decide either way at some point.

Definitely relative to current file, anything else is dangerous. Of course, absolute paths should be supported as well.

**Comment by pureconfig** [comment]
We don't want to complicate writing 1.0 parsers too much. I thus propose only supporting simple includes initially and leave partial includes/namespaces for 2.0.
